Source/missiles.cpp

```diff
@@ -2699,23 +2699,24 @@ void MI_Arrow(Missile &missile)
 {
 	missile._mirange--;
 	missile._midist++;
-	int p = missile._misource;
 
 	int mind;
 	int maxd;
-	if (!missile.IsTrap()) {
-		if (missile._micaster == TARGET_MONSTERS) {
-			const Player &player = Players[p];
-			mind = player._pIMinDam;
-			maxd = player._pIMaxDam;
-		} else {
-			auto &monster = Monsters[p];
-			mind = monster.minDamage;
-			maxd = monster.maxDamage;
-		}
-	} else {
+	switch (missile.sourceType()) {
+	case MissileSource::Player: {
+		const Player *player = missile.sourcePlayer();

```

Comment:
We already check the source type, and then `sourcePlayer/Monster` checks the type again!

Generally, for variant types, the method that returns a particular value is not checked, and the caller instead must check the type. Let's do the same?

Reply:
I think this is mostly a temporary issue until everything has been moved over.

Reply:
That means we can also change sourceMonster and sourcePlayer to return a reference instead of a pointer, right?

Reply:
Yeah